ostd/src/sync/rcu/mod.rs

```diff
@@ -2,105 +2,307 @@
     }
 
-    pub fn get(&self) -> RcuReadGuard<'_, P> {
-        let obj = unsafe { &*self.ptr.load(Acquire) };
-        RcuReadGuard { obj, rcu: self }
+    /// Retrieves a read guard for the RCU object.
+    ///
+    /// The guard allows read-only access to the data protected by RCU. If the
+    /// data is not initialized, the guard will behave as `None` before getting
+    /// dereferenced.
+    pub fn read_maybe_uninit(&self) -> RcuReadGuard<'_, P, true> {
+        let guard = disable_preempt();
+        RcuReadGuard {
+            obj_ptr: self.ptr.load(Acquire),
+            rcu: self,
+            _inner_guard: guard,
+        }
     }
 }
 
-impl<P: OwnerPtr + Send> Rcu<P> {
-    pub fn replace(&self, new_ptr: P) -> RcuReclaimer<P> {
-        let new_ptr = <P as OwnerPtr>::into_raw(new_ptr) as *mut _;
-        let old_ptr = {
-            let old_raw_ptr = self.ptr.swap(new_ptr, AcqRel);
-            unsafe { <P as OwnerPtr>::from_raw(old_raw_ptr) }
-        };
-        RcuReclaimer { ptr: old_ptr }
+impl<P: OwnerPtr + Send, const MAYBE_UNINIT: bool> Rcu_<P, MAYBE_UNINIT> {
+    /// Replaces the current pointer with a new pointer.
+    ///
+    /// This function updates the pointer to the new pointer regardless of the
+    /// original pointer. If the original pointer is not NULL, it will be
+    /// dropped after the grace period.
+    pub fn update(&self, new_ptr: P) {
+        let new_ptr = <P as OwnerPtr>::into_raw(new_ptr).cast_mut();
+        let old_raw_ptr = self.ptr.swap(new_ptr, AcqRel);
+
+        if let Some(p) = NonNull::new(old_raw_ptr) {
+            // SAFETY: It was previously returned by `into_raw`.
+            unsafe { delay_drop::<P>(p) };
+        }
     }
 }
 
-#[clippy::has_significant_drop]
-#[must_use]
-pub struct RcuReadGuard<'a, P: OwnerPtr> {
-    obj: &'a <P as OwnerPtr>::Target,
-    rcu: &'a Rcu<P>,
+/// A guard that allows read-only access to the initialized data protected
+/// by the RCU mechanism.
+pub struct RcuReadGuard<'a, P: OwnerPtr, const MAYBE_UNINIT: bool> {
+    /// If maybe uninitialized, the pointer can be NULL.
+    obj_ptr: *mut <P as OwnerPtr>::Target,
+    rcu: &'a Rcu_<P, MAYBE_UNINIT>,
+    _inner_guard: DisabledPreemptGuard,
 }
 
-impl<'a, P: OwnerPtr> Deref for RcuReadGuard<'a, P> {
+// Initialized RCU guard can be dereferenced.
+impl<P: OwnerPtr> Deref for RcuReadGuard<'_, P, false> {
     type Target = <P as OwnerPtr>::Target;
 
     fn deref(&self) -> &Self::Target {
-        self.obj
+        // SAFETY: Since the preemption is disabled, the pointer is valid
+        // because other writers won't update the pointer until this task
+        // passes the quiescent state.
+        // And this pointer is not NULL.
+        unsafe { &*self.obj_ptr }
     }
 }
 
-#[repr(transparent)]
-pub struct RcuReclaimer<P> {
-    ptr: P,
+// Maybe uninitialized RCU guard can be dereferenced after checking.
+impl<'a, P: OwnerPtr> RcuReadGuard<'a, P, true> {
+    /// Get the initialized read guard.
```

Comment:
[`Get` -> `Gets`](https://rust-lang.github.io/rfcs/0505-api-comment-conventions.html#formatting):
> The summary line should be written in third person singular present indicative form. Basically, this means write “Returns” instead of “Return”.